subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/FileSystemSnapshotterVirtualFileSystem.java

```diff
@@ -0,0 +1,115 @@
+import java.util.function.Function;
+
+public class FileSystemSnapshotterVirtualFileSystem implements VirtualFileSystem {
+
+    private final FileSystemSnapshotter snapshotter;
+    private final DefaultFileSystemMirror mirror;
+
+    public FileSystemSnapshotterVirtualFileSystem(FileHasher hasher, Interner<String> stringInterner, Stat stat, String... defaultExcludes) {
+        this.mirror = new DefaultFileSystemMirror();
+        this.snapshotter = new DefaultFileSystemSnapshotter(hasher, stringInterner, stat, mirror, defaultExcludes);
+    }
+
+    @Override
+    public void read(String location, FileSystemSnapshotVisitor visitor) {
+        maybeShortCircuitVisitor(snapshotter.snapshot(new File(location)), visitor);
+    }
+
+    @Override
+    public <T> Optional<T> readRegularFileContentHash(String location, Function<HashCode, T> visitor) {
+        HashCode hashCode = snapshotter.getRegularFileContentHash(new File(location));
+        return Optional.ofNullable(hashCode).map(visitor);
+    }
+
+    @Override
+    public void read(String location, SnapshottingFilter filter, FileSystemSnapshotVisitor visitor) {
+        FileSystemSnapshot snapshot = snapshotter.snapshotDirectoryTree(new File(location), filter);
+        if (snapshot instanceof FileSystemLocationSnapshot) {
+            maybeShortCircuitVisitor((FileSystemLocationSnapshot) snapshot, visitor);
+        } else {
+            snapshot.accept(visitor);
+        }
+    }
+
+    @Override
+    public void update(Iterable<String> locations, Runnable action) {
+        locations.forEach(location -> mirror.beforeOutputChange(locations));
+    }
+
+    @Override
+    public void invalidateAll() {
+        mirror.beforeBuildFinished();
+    }
+
+    @Override
+    public void updateWithKnownSnapshot(String location, FileSystemLocationSnapshot snapshot) {
+        determineMetadata(snapshot.getType())
+            .ifPresent(metadata -> mirror.putMetadata(snapshot.getAbsolutePath(), metadata));
+        mirror.putSnapshot(snapshot);
+    }
+
+    private Optional<FileMetadataSnapshot> determineMetadata(FileType fileType) {
+        switch (fileType) {
+            case RegularFile:
+                return Optional.empty();
+            case Directory:
+                return Optional.of(DefaultFileMetadata.directory());
+            case Missing:
+                return Optional.of(DefaultFileMetadata.missing());
+            default:
+                throw new AssertionError("Unknown file type: " + fileType);
+        }
+    }
+
+    @Override
+    public FileSystemSnapshot snapshotWithBuilder(Consumer<FileSystemSnapshotBuilder> buildAction) {
+        return snapshotter.snapshotWithBuilder(buildAction);
+    }
+
+    private void maybeShortCircuitVisitor(FileSystemLocationSnapshot snapshot, FileSystemSnapshotVisitor visitor) {
+        if (visitor instanceof MerkleDirectorySnapshotBuilder) {
+            ((MerkleDirectorySnapshotBuilder) visitor).resultKnown(snapshot);
```

Comment:
I wonder if there's a nicer way to do this.

Reply:
The method is gone now.